x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/fulltext/Match.java

```diff
@@ -393,20 +396,48 @@ public Expression replaceQueryBuilder(QueryBuilder queryBuilder) {
     public BiConsumer<LogicalPlan, Failures> postAnalysisPlanVerification() {
         return (plan, failures) -> {
             super.postAnalysisPlanVerification().accept(plan, failures);
-            plan.forEachExpression(Match.class, m -> {
-                if (m.fieldAsFieldAttribute() == null) {
+            fieldVerifier(plan, this, () -> field, failures);
+        };
+    }
+
+    public static void fieldVerifier(LogicalPlan plan, FullTextFunction function, Supplier<Expression> fieldSupplier, Failures failures) {
+        var field = fieldSupplier.get();
+        var fieldAttribute = fieldAsFieldAttribute(field);
+        if (fieldAttribute == null) {
+            plan.forEachExpression(function.getClass(), m -> {
+                if (function.children().contains(field)) {
                     failures.add(
                         Failure.fail(
-                            m.field(),
+                            field,
                             "[{}] {} cannot operate on [{}], which is not a field from an index mapping",
-                            functionName(),
-                            functionType(),
-                            m.field().sourceText()
+                            m.functionName(),
+                            m.functionType(),
+                            field.sourceText()
                         )
                     );
                 }
             });
-        };
+        } else {
+            // Traverse the plan to find the EsRelation outputting the field
+            plan.forEachDown(p -> {
+                if (p instanceof EsRelation esRelation && esRelation.indexMode() != IndexMode.STANDARD) {
+                    // Check if this EsRelation supplies the field
+                    if (esRelation.output().stream().anyMatch(attr -> attr.id().equals(fieldAttribute.id()))) {
```

Comment:
Wouldn't it be more efficient to do `esRelation.outputSet().contains(fieldAttribute)`?

Reply:
It would, ofc, thx!